src/lib.rs

```diff
@@ -56,6 +56,26 @@ where
     Err(())
 }
 
+#[cfg(target_env = "gnu")]
+mod glibc {
+    use std::path::Path;
+
+    // Attempt to compile a c program that refers to strlcpy from the std
+    // library to determine whether glibc packages it. If the function
+    // declaration cannot be found, a warning, escallated into an error,
+    // is emmited.
+    pub(crate) fn has_strlcpy() -> bool {
+        let path = Path::new(env!("CARGO_MANIFEST_DIR")).join("src/strlcpy.c");
+
+        cc::Build::new()
+            .cargo_metadata(false)
+            .warnings_into_errors(true)
+            .file(path)
+            .try_compile("has_strlcpy")
+            .is_ok()
+    }
```

Comment:
I just found the existence of [`get_compiler`](https://docs.rs/cc/latest/cc/struct.Build.html#method.get_compiler). With that we can actually link and create an executable, which is more correct. So this is better:

```rust
    pub(crate) fn has_strlcpy() -> bool {
        let c_src = Path::new(env!("CARGO_MANIFEST_DIR")).join("src/strlcpy.c");
        let bin =
            PathBuf::from(env::var("OUT_DIR").unwrap()).join("has_strlcpy");

        println!("cargo:rerun-if-changed={}", c_src.display());

        cc::Build::new()
            .warnings(false)
            .get_compiler()
            .to_command()
            .arg(c_src)
            .arg("-o")
            .arg(bin)
            .status()
            .expect("failed to execute gcc")
            .success()
    }
```

You can also revert to the older comment.

Reply:
Does it need to link, given that the CMake scripts AFAIK don't do that either?

Reply:
No we don't need to link as long as `strlcpy.c` produces only the warning we are interested in. If in the future gcc releases a new warning appears, then build will fail.

Reply:
Warning flags might even get picked up from `cc-rs`-defined environment variables, skewing the result?

CMake does a bunch of tweaks to that:
https://github.com/Kitware/CMake/blob/master/Modules/CheckSymbolExists.cmake

Reply:
They also use the address of the symbol with a cast to support generic cases (something we don't need to do now) to not have to replicate the function signature.

Reply:
> Warning flags might even get picked up from cc-rs-defined environment variables, skewing the result?

Yes. `cc-rs` is aware of `CFLAGS` and `CFLAGS_<target>`.

Reply:
I think linking is preferable to reduce the risk of false positives.